components/feature/awesomebar/src/main/java/mozilla/components/feature/awesomebar/provider/BookmarksStorageSuggestionProvider.kt

```diff
@@ -0,0 +1,49 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.awesomebar.provider
+
+import mozilla.components.browser.icons.BrowserIcons
+import mozilla.components.concept.awesomebar.AwesomeBar
+import mozilla.components.concept.storage.BookmarkNode
+import mozilla.components.concept.storage.BookmarksStorage
+import mozilla.components.feature.awesomebar.internal.loadLambda
+import mozilla.components.feature.session.SessionUseCases
+import java.util.UUID
+
+private const val BOOKMARKS_SUGGESTION_LIMIT = 20
+
+/**
+ * A [AwesomeBar.SuggestionProvider] implementation that provides suggestions based on the bookmarks
+ * stored in the [BookmarksStorage].
+ */
+class BookmarksStorageSuggestionProvider(
+    private val bookmarksStorage: BookmarksStorage,
+    private val loadUrlUseCase: SessionUseCases.LoadUrlUseCase,
+    private val icons: BrowserIcons? = null
+) : AwesomeBar.SuggestionProvider {
+
+    override val id: String = UUID.randomUUID().toString()
+
+    override suspend fun onInputChanged(text: String): List<AwesomeBar.Suggestion> {
+        if (text.isEmpty()) {
+            return emptyList()
+        }
+
+        val suggestions = bookmarksStorage.searchBookmarks(text, BOOKMARKS_SUGGESTION_LIMIT)
+        return suggestions.distinctBy { it.url }.into()
```

Comment:
Do you know in which order we get back the results? The order, ideally, needs to be stable. You can add a further `sortBy { it.guid }` here to make sure. Or, confirm with the API that the order _is_ stable across calls... But best to make sure on our end.

Reply:
From api docs: "The order of the results is unspecified." - so yeah, sort this.

Reply:
cc @thomcc if you feel like changing the API :)

Reply:
Yep, totally willing to change how that function works to better suit usage.